File: python/29. Divide Two Integers.py
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        sign = 1
        n1, n2 = dividend, divisor
        if n1 == 0 or abs(n1) < abs(n2):
            return 0
        if (n1 > 0 and n2 < 0) or (n1 < 0 and n2 > 0):
            sign = 0
        n1, n2 = abs(n1), abs(n2)
        res = 0
        while n1 >= n2:
            t = 1
            b = n2
            while n1 > b<<1:
                b = b<<1
                t = t<<1
            res += t
            n1 -= b
        res = sign and res or -res
        if res > pow(2,31)-1 or res < -pow(2,31):
            return pow(2,31)
        return res
            
            
s = Solution()
print s.divide(3, -2)        

File: go/src/leetcode/29. Divide Two Integers/29. Divide Two Integers.go
package main

import (
	"fmt"
	"math"
)

func divide(dividend int, divisor int) int {
	if dividend == 0 || abs(dividend) < abs(divisor) {
		return 0
	}
	sign := 1
	if (dividend > 0 && divisor < 0) || (dividend < 0 && divisor > 0) {
		sign = -1
	}
	ans := int64(0)
	a, b := abs(dividend), abs(divisor)
	for a >= b {
		k := int64(1)
		t := b
		for a > t<<1 {
			t <<= 1
			k <<= 1
		}
		ans += k
		a -= t
	}

	tmp := int64(sign) * ans
	if tmp > math.MaxInt32 || tmp < math.MinInt32 {
		return math.MaxInt32
	}
	return int(tmp)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	fmt.Println(divide(10, 3))
	fmt.Println(divide(7, -3))
}
